V compiler bug

### Describe the bug

error is
```
================== C compilation error (from tcc): ==============
cc: /tmp/v_1000/BitHavoc.01KBSXMB9V9R2STYJ1VDXW2G24.tmp.c:6447: error: invalid operand types for binary operation
=================================================================
Try passing `-g` when compiling, to see a .v file:line information, that correlates more with the C error.
(Alternatively, pass `-show-c-output`, to print the full C error message).
builder error: 
==================
C error found. It should never happen, when compiling pure V code.
This is a V compiler bug, please report it using `v bug file.v`,
or goto https://github.com/vlang/v/issues/new/choose .
You can also use #help on Discord: https://discord.gg/vlang .
```
when compiling pure C code

### Reproduction Steps

parser.v
``` v
module parser

// main parsor for the BitHovoc parser

struct Token {
  type string
	token string
}

pub fn parse(code string) []Token{
  //loop over letters

  len_byte := 8
  len_line := 5
  mut current_char := 0
  mut current_line := 0
  mut current_byte := 0
  mut tokens := []Token{}

  for mut letter in code.runes() {
    if letter.str() == "/n" || current_byte >= len_line {
      current_line++
      current_char = 0
      current_byte = 0
      tokens << Token{type: "NEWLINE", token: letter.str() }
    }

    if letter == " " {
      current_byte++
      current_char = 0
      tokens << Token{type: "SPLIT", token: " " }
    }

    if current_char >= len_byte {
      current_byte++
      current_char = 0
      tokens << Token{type: "SPLIT", token: " " }
    }
  }
  return tokens
}
```
main.v
```v
module main
import parser

fn main() {
	println("this is a BitHavok... The language that likes binary")
	println(parser.parse("00000001 00000001 10000101 00000001 00000001 00000001"))
}
```

### Expected Behavior

I expected the code to compile and run the program and print the token stream from the tokenise function

### Current Behavior

on a `v run .`
```
================== C compilation error (from tcc): ==============
cc: /tmp/v_1000/BitHavoc.01KBSXMB9V9R2STYJ1VDXW2G24.tmp.c:6447: error: invalid operand types for binary operation
=================================================================
Try passing `-g` when compiling, to see a .v file:line information, that correlates more with the C error.
(Alternatively, pass `-show-c-output`, to print the full C error message).
builder error: 
==================
C error found. It should never happen, when compiling pure V code.
This is a V compiler bug, please report it using `v bug file.v`,
or goto https://github.com/vlang/v/issues/new/choose .
You can also use #help on Discord: https://discord.gg/vlang .
```
### Possible Solution

_No response_

### Additional Information/Context

_No response_

### V version

V 0.4.12 f108077

### Environment details (OS name and version, etc.)
```
|V full version      |V 0.4.12 58477f3dc362bbcde6199d1cfb853a0af356d721
|:-------------------|:-------------------
|OS                  |linux, "Garuda Linux Broadwing"
|Processor           |4 cpus, 64bit, little endian, Intel(R) Core(TM) i7-3517U CPU @ 1.90GHz
|Memory              |0.27GB/3.7GB
|                    |
|V executable        |/home/darionk/v/v
|V last modified time|2025-12-02 18:45:38
|                    |
|V home dir          |OK, value: /home/darionk/v
|VMODULES            |OK, value: /home/darionk/.vmodules
|VTMP                |OK, value: /tmp/v_1000
|Current working dir |OK, value: /home/darionk/Code/v/BitHavoc
|                    |
|Git version         |git version 2.52.0
|V git status        |58477f3d
|.git/config present |true
|                    |
|cc version          |cc (GCC) 15.2.1 20251112
|gcc version         |gcc (GCC) 15.2.1 20251112
|clang version       |N/A
|tcc version         |tcc version 0.9.28rc 2025-02-13 HEAD@f8bd136d (x86_64 Linux)
|tcc git status      |thirdparty-linux-amd64 696c1d84
|emcc version        |N/A
|glibc version       |ldd (GNU libc) 2.42
```
> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

--- a/vlib/v/checker/infix.v
+++ b/vlib/v/checker/infix.v
@@ -868,9 +868,26 @@ fn (mut c Checker) infix_expr(mut node ast.InfixExpr) ast.Type {
 			right_type = c.unwrap_generic(right_type)
 			right_sym = c.table.sym(right_type)
 		}
-		if !(c.symmetric_check(left_type, right_type) && c.symmetric_check(right_type, left_type))
-			&& !c.pref.translated && !c.file.is_translated && !node.left.is_auto_deref_var()
-			&& !node.right.is_auto_deref_var() {
+		types_match := c.symmetric_check(left_type, right_type)
+			&& c.symmetric_check(right_type, left_type)
+		mut types_match_after_deref := false
+		mut deref_left_type := left_type
+		mut deref_right_type := right_type
+		if !types_match && (node.left.is_auto_deref_var() || node.right.is_auto_deref_var()) {
+			deref_left_type = if node.left.is_auto_deref_var() {
+				left_type.deref()
+			} else {
+				left_type
+			}
+			deref_right_type = if node.right.is_auto_deref_var() {
+				right_type.deref()
+			} else {
+				right_type
+			}
+			types_match_after_deref = c.symmetric_check(deref_left_type, deref_right_type)
+				&& c.symmetric_check(deref_right_type, deref_left_type)
+		}
+		if !types_match && !types_match_after_deref && !c.pref.translated && !c.file.is_translated {
 			// for type-unresolved consts
 			if left_type == ast.void_type || right_type == ast.void_type {
 				return ast.void_type
@@ -883,7 +900,17 @@ fn (mut c Checker) infix_expr(mut node ast.InfixExpr) ast.Type {
 			if node.right is ast.None && left_is_option {
 				return ast.bool_type
 			}
-			c.error('infix expr: cannot use `${right_sym.name}` (right expression) as `${left_sym.name}`',
+			error_left_sym := if node.left.is_auto_deref_var() {
+				c.table.sym(deref_left_type)
+			} else {
+				left_sym
+			}
+			error_right_sym := if node.right.is_auto_deref_var() {
+				c.table.sym(deref_right_type)
+			} else {
+				right_sym
+			}
+			c.error('infix expr: cannot use `${error_right_sym.name}` (right expression) as `${error_left_sym.name}`',
 				left_right_pos)
 		} else if left_type.is_ptr() {
 			for_ptr_op := c.table.type_is_for_pointer_arithmetic(left_type)

--- a/vlib/v/checker/tests/for_mut_compare_rune_string_err.out
+++ b/vlib/v/checker/tests/for_mut_compare_rune_string_err.out
@@ -1,0 +1,7 @@
+vlib/v/checker/tests/for_mut_compare_rune_string_err.vv:4:6: error: infix expr: cannot use `string` (right expression) as `rune`
+    2 |     s := 'Hello, World!'
+    3 |     for mut c in s.runes() {
+    4 |         if c == ' ' {
+      |            ~~~~~~~~
+    5 |             c = `x`
+    6 |         }

--- a/vlib/v/checker/tests/for_mut_compare_rune_string_err.vv
+++ b/vlib/v/checker/tests/for_mut_compare_rune_string_err.vv
@@ -1,0 +1,8 @@
+fn main() {
+	s := 'Hello, World!'
+	for mut c in s.runes() {
+		if c == ' ' {
+			c = `x`
+		}
+	}
+}

